core/render_management.ts

```diff
@@ -58,14 +70,29 @@ function doRenders() {
  * @param block The block to rerender.
  */
 function renderBlock(block: BlockSvg) {
+  if (!dirtyBlocks.has(block)) return;
   for (const child of block.getChildren(false)) {
     renderBlock(child);
   }
-  if (dirtyBlocks.has(block)) {
-    dirtyBlocks.delete(block);
-    rootBlocks.delete(block);
-    block.render(false);
-  } else {
-    block.updateConnectionLocations();
+  block.renderEfficiently();
+}
+
+/**
+ * Updates the connection database with the new locations of all of the
+ * connections that are children of the given block.
+ *
+ * @param block The block to update the connection locations of.
+ * @param blockTL The top left of the given block in workspace coordinates.
+ */
+function updateConnectionLocations(block: BlockSvg, blockTL: Coordinate) {
```

Comment:
nit: would prefer a different name like "location" -- the fact that it's the top left of the block isn't particularly relevant here and `blockTL` is a fairly inscrutable name imo

Reply:
How do you feel about `blockOrigin`?

Personally I think the fact that it's the top left of the block is pretty important, but that may just be a me thing. Knowing that it's at the top left helps me visualize how all of the offsets in the positioning logic add together.

Reply:
sgtm.

What I mean by not relevant is that if this coordinate were relative to the bottom right of the block instead, as long as all the coordinates were using the same relative corner, nothing in this method would change. It would be somewhere further down in the calls that cares about which specific corner we are looking at, in terms of bumping and positioning, that cares which corner it is. but in any case, `blockOrigin` seems good regardless :)